common/js/src/executable-module/offscreen-doc-emscripten-module.js

```diff
@@ -0,0 +1,217 @@
+    /** @type {boolean} @private */
+    this.offscreenDocActive_ = false;
+    /** @type {!GSC.DelayedMessageChannel} @const @private */
+    this.messageChannel_ = new GSC.DelayedMessageChannel();
+    /** @type {!GSC.PortMessageChannel|null} @private */
+    this.statusChannel_ = null;
+    /** @type {!goog.promise.Resolver<void>} @const @private */
+    this.loadPromiseResolver_ = goog.Promise.withResolver();
+    GSC.PromiseHelpers.suppressUnhandledRejectionError(
+        this.loadPromiseResolver_.promise);
+  }
+
+  /** @override */
+  getLogger() {
+    return this.logger_;
+  }
+
+  /** @override */
+  startLoading() {
+    // Start the asynchronous load operation. Tie the load promise with its
+    // result.
+    this.load_().then(
+        () => this.loadPromiseResolver_.resolve(),
+        (error) => this.loadPromiseResolver_.reject(error));
+  }
+
+  /** @override */
+  getLoadPromise() {
+    return this.loadPromiseResolver_.promise;
+  }
+
+  /** @override */
+  getMessageChannel() {
+    return this.messageChannel_;
+  }
+
+  /** @override */
+  disposeInternal() {
+    if (this.offscreenDocActive_)
+      chrome.offscreen.closeDocument();
+    this.messageChannel_.dispose();
+    if (this.statusChannel_)
+      this.statusChannel_.dispose();
+    this.loadPromiseResolver_.reject();
+    super.disposeInternal();
+  }
+
+  /**
+   * @returns {!Promise<void>}
+   * @private
+   */
+  async load_() {
+    // Listen for the incoming message channel from the Offscreen Document.
+    const statusChannelWaiter =
+        new GSC.PortMessageChannelWaiter(STATUS_PORT_NAME);
+
+    // Prepare parameters for the Offscreen Document. The module name is passed
+    // via URL query parameters.
+    const url = new URL(OFFSCREEN_DOC_URL, globalThis.location.href);
+    url.searchParams.append(URL_PARAM, this.moduleName_);
+    const docParams = {
+      'url': url.toString(),
+      'reasons': OFFSCREEN_DOC_REASONS,
+      'justification': OFFSCREEN_DOC_JUSTIFICATION,
+    };
+
+    // Load the Offscreen Document.
+    await chrome.offscreen.createDocument(docParams);
+    if (this.isDisposed()) {
+      // If we got disposed of while waiting for the
+      // `chrome.offscreen.createDocument()` completion, the logic in
+      // `disposeInternal()` didn't have a chance to clean up the document.
+      chrome.offscreen.closeDocument();
+      throw new Error('Disposed');
+    }
+    this.offscreenDocActive_ = true;
+
+    // Wait until the Offscreen Document opens the status message channel.
+    const statusChannel = await statusChannelWaiter.getPromise();
+    if (this.isDisposed()) {
+      statusChannel.dispose();
+      throw new Error('Disposed');
+    }
+    this.statusChannel_ = statusChannel;
+
+    // Wait until the Offscreen Document signals us about the module's
+    // loaded/disposed status (whichever comes first).
+    await Promise.all([this.waitLoadedMessage_(), this.waitDisposedMessage_()]);
```

Comment:
The way I read the documentation, this wait on all promises to be resolved or one to be rejected. I think this should be Promise.race instead.

Reply:
Oh, a great catch!